src/game/server/tf/tf_player.cpp

```diff
@@ -15146,12 +15154,37 @@ void CTFPlayer::PainSound( const CTakeDamageInfo &info )
 			TFPlayerClassData_t *pData = GetPlayerClass()->GetData();
 			if ( pData )
 			{
-				EmitSound( pData->GetDeathSound( DEATH_SOUND_GENERIC ) );
+				CPASFilter filter( GetAbsOrigin() );
+				if ( m_Shared.InCond( TF_COND_DISGUISED ) )
+				{
+					filter.RemoveRecipientsByTeam( GetGlobalTFTeam( ( GetTeamNumber() == TF_TEAM_RED ) ? TF_TEAM_BLUE : TF_TEAM_RED ) );
+					filter.RemoveRecipient( this );
+				}
+
+				EmitSound( filter, entindex(), pData->GetDeathSound( DEATH_SOUND_GENERIC ) );
+			}
+
+			if ( m_Shared.InCond( TF_COND_DISGUISED ) )
+			{
+				TFPlayerClassData_t *pDisguiseData = GetPlayerClassData( m_Shared.GetDisguiseClass() );
+				if ( pDisguiseData )
+				{
+					CPASFilter disguisedFilter( GetAbsOrigin() );
+					disguisedFilter.RemoveRecipientsByTeam( GetGlobalTeam( TEAM_SPECTATOR ) );
+					disguisedFilter.RemoveRecipientsByTeam( GetGlobalTeam( TEAM_UNASSIGNED ) );
+					disguisedFilter.RemoveRecipientsByTeam( GetGlobalTFTeam( GetTeamNumber() ) );
+					disguisedFilter.AddRecipient( this );
+					EmitSound( disguisedFilter, entindex(), pDisguiseData->GetDeathSound( DEATH_SOUND_GENERIC ) );
+				}
 			}
 		}
 		return;
 	}
 
+	// no pain sounds while disguised, our man has supreme discipline
```

Comment:
why does this check happen twice? Would it be better (or at least more clear) to instead have a flag "we already played pain noise, skip the remaining logic" boolean field?

Reply:
If the damage received was DMG_FALL it already returns early. The idea behind checking it a 2nd time was to catch if it was not DMG_FALL, since disguised spies shouldn't emit anything then.

Reply:
You might be able to fold it into the DMG_FALL check then:
```
	// fall damage and disguise fall damage enabled OR not disguised
	if ( info.GetDamageType() & DMG_FALL && ( !m_Shared.InCond( TF_COND_DISGUISED ) || tf_disguise_mimic_fall_damage.GetBool() ) )
	{
```